test/withStyles_test.jsx

```diff
@@ -430,6 +430,563 @@ describe('withStyles()', () => {
+      )).to.throw();
+
+      expect(() => render(
+        <FailingFontSizeComponent />,
+      )).to.throw();
+
+      expect(() => render(
+        <ExtendedComponent />,
+      )).to.not.throw();
+    });
+
+    it('validates all extending styles across all classNames against the defined predicates', () => {
+      function MyComponent() {
+        return null;
+      }
+
+      const WrappedComponent = withStyles(
+        () => ({
+          container: {
+            background: 'red',
+            color: 'blue',
+          },
+          innerContainer: {
+            fontSize: 12,
+          },
+        }),
+        {
+          extendableStyles: {
+            container: {
+              background: (v, t) => (v === t.color.red),
+              color: v => (v === 'green'),
+            },
+            innerContainer: {
+              fontSize: v => (v === 12),
+            },
+          },
+        },
+      )(MyComponent);
+      const FailingColorComponent = WrappedComponent.extendStyles(() => ({
+        container: {
+          background: 'purple', // fails
+          color: 'green',
+        },
+        innerContainer: {
+          fontSize: 12,
+        },
+      }));
+      const FailingBackgroundComponent = WrappedComponent.extendStyles(theme => ({
+        container: {
+          background: theme.color.red,
+          color: 15, // fails
+        },
+        innerContainer: {
+          fontSize: 12,
+        },
+      }));
+      const FailingFontSizeComponent = WrappedComponent.extendStyles(theme => ({
+        container: {
+          background: theme.color.red,
+          color: 'green',
+          fontSize: '12',
+        },
+        innerContainer: {
+          fontSize: '12', // fails
+        },
+      }));
+      const ExtendedComponent = WrappedComponent.extendStyles(theme => ({
+        container: {
+          background: theme.color.red,
+          color: 'green',
+        },
+        innerContainer: {
+          fontSize: 12,
+        },
+      }));
+
+      expect(() => render(
+        <FailingColorComponent />,
+      )).to.throw();
+
+      expect(() => render(
+        <FailingBackgroundComponent />,
+      )).to.throw();
+
+      expect(() => render(
+        <FailingFontSizeComponent />,
+      )).to.throw();
+
+      expect(() => render(
+        <ExtendedComponent />,
+      )).to.not.throw();
+    });
+
+    it('does not attach an "extendStyles" static property when "extendableStyles" option is not defined', () => {
+      function MyComponent() {
+        return null;
+      }
+
+      const WrappedComponent = withStyles(
+        () => ({
+          container: {
+            background: 'red',
+            color: 'blue',
+          },
+          innerContainer: {
+            fontSize: 12,
+          },
+        }),
+        {}, // extendableStyles is not defined
+      )(MyComponent);
+
+      expect(WrappedComponent.extendStyles).to.equal(undefined);
+    });
+  });
```

Comment:
Similar to https://github.com/airbnb/eslint-plugin-react-with-styles/pull/37#discussion_r269680246

Can we add some valid and invalid test cases for nested styles? Media queries, `@supports` queries, pseudo-elements, pseudo-selectors, etc.